File: clients/algoliasearch-client-kotlin/client/src/commonMain/kotlin/com/algolia/client/transport/RequestOptions.kt
package com.algolia.client.transport

import kotlinx.serialization.json.JsonObject
import kotlin.time.Duration

/**
 * Represents options for configuring a request to an endpoint.
 *
 * @property writeTimeout The write timeout for the request in milliseconds.
 * @property readTimeout The read timeout for the request in milliseconds.
 * @property headers A mutable map of header names to their respective values to be sent with the request.
 * @property urlParameters A mutable map of URL parameter names to their respective values to be appended to the request URL.
 * @property body A JSON object representing the request body.
 */
public data class RequestOptions(
  public val writeTimeout: Duration? = null,
  public val readTimeout: Duration? = null,
  public val headers: Map<String, Any> = emptyMap(),
  public val urlParameters: Map<String, Any> = emptyMap(),
  public val body: JsonObject? = null,
)

File: clients/algoliasearch-client-dart/packages/client_core/lib/src/transport/request_options.dart
/// Represents options for configuring a request to an endpoint.
final class RequestOptions {
  ///  The write timeout for the request in milliseconds.
  final Duration? writeTimeout;

  /// The read timeout for the request in milliseconds.
  final Duration? readTimeout;

  /// Header names to their respective values to be sent with the request.
  final Map<String, dynamic> headers;

  /// URL parameter names to their respective values to be appended to the request URL.
  final Map<String, dynamic> urlParameters;

  /// Custom request body.
  final dynamic body;

  const RequestOptions({
    this.writeTimeout,
    this.readTimeout,
    this.headers = const {},
    this.urlParameters = const {},
    this.body,
  });

  @override
  bool operator ==(Object other) =>
      identical(this, other) ||
      other is RequestOptions &&
          runtimeType == other.runtimeType &&
          writeTimeout == other.writeTimeout &&
          readTimeout == other.readTimeout &&
          headers == other.headers &&
          urlParameters == other.urlParameters &&
          body == other.body;

  @override
  int get hashCode =>
      writeTimeout.hashCode ^
      readTimeout.hashCode ^
      headers.hashCode ^
      urlParameters.hashCode ^
      body.hashCode;

  @override
  String toString() {
    return 'RequestOptions{writeTimeout: $writeTimeout, readTimeout: $readTimeout, headers: $headers, urlParameters: $urlParameters, body: $body}';
  }
}
